### PR TITLE
GET and PUT volunteer endpoints + frontend volunteer profile

### PR DESCRIPTION
**Pull request by jansm04**
implemented GET endpoints to get a volunteer by ID and to get all volunteers in the volunteers table
implemented PUT endpoint to update a volunteer in the volunteers table
created volunteer profile page on frontend

### CODE REVIEW

File: backend/src/routes/volunteerRoutes.ts [hunk 1]
```diff
@@ -1 +1,18 @@
...
+const router = Router();
+
+// get volunteer profile by id
+router.get('/volunteer/:volunteer_id', (req: Request, res: Response) => { getVolunteerById(req, res) });
+
+// get all volunteers
+router.get('/volunteers', (req: Request, res: Response) => { getVolunteers(req, res) });
+
+// update volunteer profile
+router.put('/volunteer/:volunteer_id', (req: Request, res: Response) => { updateVolunteer(req, res) });
```
Comment (jjessieshang) [inline backend/src/routes/volunteerRoutes.ts:18]: need `export default router` at the bottom of this file so we can define it in the index.ts file

Comment (jjessieshang) [reply backend/src/routes/volunteerRoutes.ts:18]: Then in src/index.ts we will need to define a parent path to all volunteer routes like so:

import volunteerRouter from './routes/volunteerRoutes.js';
...
app.use('/volunteer', volunteerRouter);

so all volunteer endpoints will have this parent path

File: backend/src/controllers/volunteerController.ts [hunk 2]
```diff
@@ -1 +1,61 @@
...
+            error: "Missing required parameter: 'volunteer_id'"
+        });
+    }
+
+    try {
+        const volunteer = await volunteerModel.getVolunteerById(volunteer_id);
+        res.status(200).json(volunteer);
+    } catch (error) {
+        return res.status(500).json({
+            error: "Internal server error"
```
Comment (jjessieshang) [inline backend/src/controllers/volunteerController.ts:20]: When testing the error case here, the custom error messages thrown by VolunteerModel.js were not re thrown here (by the controller functions),  they were overwritten by the "Internal Server Error" messages. Could change the error handling to:

`        return res.status(500).json({
            error: `Internal server error: ${error}`
        });`

to show the error returned from the inner function calls.

File: backend/src/models/volunteerModel.ts [hunk 3]
```diff
@@ -1 +1,49 @@
...
+    updateVolunteer(volunteer_id: string, volunteerData: any): Promise<any> {
+        return new Promise((resolve, reject) => {
+            // Construct the SET clause dynamically
+            const setClause = Object.keys(volunteerData).map(key => `${key} = ?`).join(', ');
+            const query = `UPDATE volunteers SET ${setClause} WHERE volunteer_id = ?`;
+            const values = [...Object.values(volunteerData), volunteer_id];
+
+            connectionPool.query(query, values, (error: any, results: any) => {
+                if (error) {
+                    return reject(error);
```
Comment (jjessieshang) [inline backend/src/models/volunteerModel.ts:43]: specific error msg here

File: frontend/src/components/UserProfileForm.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        const data = await response.json();
+        setUserData(data);
+      } catch (error) {
+        console.error('Error fetching user data:', error);
+      }
+    };
+
+    const fetchVolunteerData = async () => {
+      try {
+        const response = await fetch('http://localhost:3001/api/sudo_get_VOLUN?user_id=' + userData.user_id);
```
Comment (jjessieshang) [inline frontend/src/components/UserProfileForm.js:37]: make the api call in a volunteerService.js file in frontend/src/api , then call the function here

**Review by jjessieshang — COMMENTED** [review]
Structure is great just some notes!

File: backend/src/routes/volunteerRoutes.ts (hunk 1 above)
Comment (jansm04) [reply backend/src/routes/volunteerRoutes.ts:18]: sure np

File: backend/src/controllers/volunteerController.ts (hunk 2 above)
Comment (jansm04) [reply backend/src/controllers/volunteerController.ts:20]: sure np

**Review by Mageacallll — APPROVED** [review]
It seems to be an update for better interpretability! Great improvement and I believe it is good to go. (Still need to do some local test based on localized small SQLite mock database later, will do it around the weekend.)

**Comment by jjessieshang** [comment]
Worked with Bryan to place the edit profile form component in the "my profile" tab. Also altered the volunteer table schema to include the email field for ease of querying.

**Review by JoshFung — APPROVED** [review]
Looks good to me!

**Pull request merged**